File: testSuite/test-checkpointing.pl
#!/usr/bin/env perl
use strict;
use warnings;
use PDL;
use PDL::IO::HDF5;
use PDL::NiceSlice;

# Run a set of Galacticus models to test checkpointing functionality.
# Andrew Benson (05-July-2023)

# Ensure output directory exists.
system("mkdir -p outputs");

# # Run model without checkpointing.
system("export OMP_NUM_THREADS=1; cd ..; ./Galacticus.exe testSuite/parameters/checkpointingNoCheckpoints.xml"  );
die("FAILED: failed to run model without checkpointing")
    unless ( $? == 0 );

# # Run model with checkpointing - interrupted.
system("export OMP_NUM_THREADS=1; cd ..; rm -f checkpoint.chk; ./Galacticus.exe testSuite/parameters/checkpointingCheckpoints.xml");
die("FAILED: failed to produce a checkpoint file")
    unless ( -e "../checkpoint.chk" );

# # Run model with checkpointing - resuming.
system("export OMP_NUM_THREADS=1; cd ..; ./Galacticus.exe testSuite/parameters/checkpointingResume.xml");
die("FAILED: failed to run model resuming from checkpoint")
    unless ( $? == 0 );

# Read model data.
my $data;
foreach my $modelName ( "chechkpointingNoCheckpoints", "chechkpointingCheckpoints" ) {
    my $model = new PDL::IO::HDF5("outputs/".$modelName.".hdf5");
    my $nodes = $model->group("Outputs/Output1/nodeData");
    $data->{$modelName}->{$_} = $nodes->dataset($_)->get()
	foreach ( "nodeIndex", "basicMass" );
}

# Compare results.
my $status = "SUCCESS";
for(my $i=0;$i<nelem($data->{'chechkpointingNoCheckpoints'}->{'nodeIndex'});++$i) {
    my $match = which($data->{'chechkpointingCheckpoints'}->{'nodeIndex'} == $data->{'chechkpointingNoCheckpoints'}->{'nodeIndex'}->(($i)));
    next
	unless ( nelem($match) == 1 );
    my $massNoCheckpoints = $data->{'chechkpointingNoCheckpoints'}->{'basicMass'}          ->(($i));
    my $massCheckpoints   = $data->{'chechkpointingCheckpoints'  }->{'basicMass'}->($match)->(( 0));
    $status = "FAIL"
	unless ( $massNoCheckpoints == $massCheckpoints );
}
print $status.": resume from chekpoint file\n";

exit;

File: testSuite/test-checkpointing.py
#!/usr/bin/env python3
import subprocess
import sys
import os
import h5py
import numpy as np

# Run a set of Galacticus models to test checkpointing functionality.
# Andrew Benson (ported to Python)

# Ensure output directory exists.
subprocess.run("mkdir -p outputs", shell=True)

# Run model without checkpointing.
status = subprocess.run("export OMP_NUM_THREADS=1; cd ..; ./Galacticus.exe testSuite/parameters/checkpointingNoCheckpoints.xml", shell=True)
if status.returncode != 0:
    print("FAILED: failed to run model without checkpointing")
    sys.exit(0)

# Run model with checkpointing - interrupted.
subprocess.run("cd ..; rm -f checkpoint.chk", shell=True)
subprocess.run("export OMP_NUM_THREADS=1; cd ..; ./Galacticus.exe testSuite/parameters/checkpointingCheckpoints.xml", shell=True)
if not os.path.exists("../checkpoint.chk"):
    print("FAILED: failed to produce a checkpoint file")
    sys.exit(0)

# Run model with checkpointing - resuming.
status = subprocess.run("export OMP_NUM_THREADS=1; cd ..; ./Galacticus.exe testSuite/parameters/checkpointingResume.xml", shell=True)
if status.returncode != 0:
    print("FAILED: failed to run model resuming from checkpoint")
    sys.exit(0)

# Read model data.
data = {}
for modelName in ("checkpointingNoCheckpoints", "checkpointingCheckpoints"):
    with h5py.File(f"outputs/{modelName}.hdf5", "r") as model:
        nodes = model["Outputs/Output1/nodeData"]
        data[modelName] = {
            "nodeIndex": nodes["nodeIndex"][:],
            "basicMass": nodes["basicMass"][:],
        }

# Compare results.
status_str        = "SUCCESS"
toleranceRelative = 0.01
noChk = data["checkpointingNoCheckpoints"]
chk   = data["checkpointingCheckpoints"]
for i in range(len(noChk["nodeIndex"])):
    match = np.where(chk["nodeIndex"] == noChk["nodeIndex"][i])[0]
    if len(match) != 1:
        continue
    massNoCheckpoints = noChk["basicMass"][i]
    massCheckpoints   = chk["basicMass"][match[0]]
    agrees            = abs(massNoCheckpoints - massCheckpoints) < toleranceRelative * massNoCheckpoints
    if not agrees:
        status_str = "FAIL"
        print(f"\t({i}) {massNoCheckpoints} == {massCheckpoints} ? : FAILURE")

print(f"{status_str}: resume from checkpoint file")
